libraries/chain/controller.cpp

```diff
@@ -3999,15 +3999,35 @@ struct controller_impl {
       bsp->verify_qc(qc_proof);
    }
 
+   // Verify a Legacy block does not have a QC block extension.
+   // (It can have a finality block header extension during transition to
+   // Savanna)
+   void verify_legacy_block_exts( const signed_block_ptr& b ) {
+      uint32_t block_num = b->block_num();
+
+      auto block_exts = b->validate_and_extract_extensions();
+      auto qc_ext_id = quorum_certificate_extension::extension_id();
+      bool qc_extension_present = block_exts.count(qc_ext_id) != 0;
+      EOS_ASSERT( !qc_extension_present,
+                  block_validate_exception,
+                  "Legacy block #${b} includes a QC block extension",
+                  ("b", block_num) );
```

Comment:
Seems like this should verify that that `qc_claim` is as expected.
If present then `qc_claim.is_qc_strong` must be `false`.
If present and prev is present then their `qc_claim` block numbers are equal.
If prev not present then `qc_claim` block number is this block number and contains a `new_finalizer_policy_diff`.

Reply:
Looking good. Add check for the new_finalizer_policy_diff.

Reply:
> Add check for the new_finalizer_policy_diff.

Good call.

If I am not mistaken, the only protection we have against a malformed Savanna Genesis block that does not have a `new_finalizer_policy_diff` present is this `assert`:
https://github.com/AntelopeIO/spring/blob/31e94812b92acf31717ca0a3b11db82b77888773/libraries/chain/block_state.cpp#L71

But that only runs in debug mode. If it wasn't present in a Savanna Genesis block, then I believe in a release build of nodeos we would get undefined behavior in `create_if_genesis_block`.

Reply:
I think it should also attempt to apply the `new_finalizer_policy_diff` and verify `generation` is 1.
Also verify that the `new_proposer_policy_diff` is null for all `qc_claim`s. The finalizer_policy apply_diff actually needs some checks as well. I'll get a PR up for that.

Reply:
At the time when `verify_legacy_block_exts()` is called, we only have previous `block_header_state_legacy` available. We don't have access to previous proposed finalizer policy which is needed to apply_diff.

Reply:
But I will do a check on Genesis block and apply a diff to make sure its generation is 1.

Reply:
The diff would be against an empty container. Just do this:
```
finalizer_policy{}.apply_diff(std::move(*f_ext.new_finalizer_policy_diff));
```